[Cocoa] Add text extraction support for internal Swift clients

whsieh opened the pull request:
#### 6d2dce3223e6d0cba79b5428703516fd0647566b
<pre>
[Cocoa] Add text extraction support for internal Swift clients
<a href="https://bugs.webkit.org/show_bug.cgi?id=269207">https://bugs.webkit.org/show_bug.cgi?id=269207</a>
<a href="https://rdar.apple.com/121132700">rdar://121132700</a>

Reviewed by Aditya Keerthi, Abrar Rahman Protyasha and Elliott Williams.

Make several adjustments to support for text extraction; see below for more details.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add a new preference to guard runtime enablement of text extraction.

* Source/WebKit/Configurations/WebKitSwift.xcconfig:

Exclude `WKWebView+TextExtraction.swift` when using non-internal SDKs. Also, add the Swift frontend
flag `-experimental-spi-only-imports`, so that we can use `@_spiOnly`.

* Source/WebKit/DerivedSources-input.xcfilelist:

Add `WKWebView+TextExtraction.swift.in` to the derived sources input file list.

* Source/WebKit/DerivedSources-output.xcfilelist:

Add `WKWebView+TextExtraction.swift` to the derived sources output file list.

* Source/WebKit/DerivedSources.make:

Add a DerivedSources make rule for `WKWebView+TextExtraction.swift` (copied from
`WKWebView+TextExtraction.swift.in`, in WebKitAdditions).

* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setTextExtractionEnabled:]):
(-[WKPreferences _textExtractionEnabled]):

Add an SPI flag for internal clients to enable or disable text extraction.

* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):

Add a new category that implements an ObjC selector, `-_requestTextExtraction:completionHandler:`,
which is invoked via `.perform(Selector((…)))` from `WKWebView+TextExtraction.swift`.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm:
(WebKit::prepareTextExtractionSupport):

Add a helper method to ensure that `libWebKitSwift` is loaded into memory; called when an internal
client enables text extraction via the new preference.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Add the new `WKWebView+TextExtraction.swift` derived source to WebKitSwift.

Canonical link: <a href="https://commits.webkit.org/274511@main">https://commits.webkit.org/274511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/433f58bd5f74bd924abe87175e73a970879d4e74

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39276 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41629 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41810 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35176 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21113 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15584 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32857 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39850 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15366 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34041 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13351 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13319 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34981 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43088 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/32729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35661 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35310 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39125 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/38902 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11627 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37364 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/39323 "Build is in progress. Recent messages:") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15694 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/45908 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8792 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15357 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9361 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder reviewed (approved):
Code changes look good to me. May want someone else to review the `xcconfig` changes.

aprotyas reviewed (commented):
LGTM

emw-apple commented on Source/WebKit/Configurations/WebKitSwift.xcconfig:
```diff
@@ -58,3 +61,16 @@ PRIVATE_HEADERS_FOLDER_PATH = $(WEBKITSWIFT_HEADERS_FOLDER_PATH);
+SPI_ONLY_IMPORTS_FLAGS_appletvos = -Xfrontend -experimental-spi-only-imports
+SPI_ONLY_IMPORTS_FLAGS_appletvsimulator = -Xfrontend -experimental-spi-only-imports
+SPI_ONLY_IMPORTS_FLAGS_watchos = -Xfrontend -experimental-spi-only-imports
+SPI_ONLY_IMPORTS_FLAGS_watchsimulator = -Xfrontend -experimental-spi-only-imports
```
I think the way I would express this is with an sdk condition:

```
SPI_ONLY_IMPORTS_FLAGS = -Xfrontend -experimental-spi-only-imports;
SPI_ONLY_IMPORTS_FLAGS[sdk=macosx11*] = ;
SPI_ONLY_IMPORTS_FLAGS[sdk=macosx12*] = ;
```

whsieh replied:
Good call!

emw-apple reviewed (commented):
r=me for build system changes

whsieh committed:
[Cocoa] Add text extraction support for internal Swift clients
https://bugs.webkit.org/show_bug.cgi?id=269207
rdar://121132700

Reviewed by Aditya Keerthi, Abrar Rahman Protyasha and Elliott Williams.

Make several adjustments to support for text extraction; see below for more details.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add a new preference to guard runtime enablement of text extraction.

* Source/WebKit/Configurations/WebKitSwift.xcconfig:

Exclude `WKWebView+TextExtraction.swift` when using non-internal SDKs. Also, add the Swift frontend
flag `-experimental-spi-only-imports`, so that we can use `@_spiOnly`.

* Source/WebKit/DerivedSources-input.xcfilelist:

Add `WKWebView+TextExtraction.swift.in` to the derived sources input file list.

* Source/WebKit/DerivedSources-output.xcfilelist:

Add `WKWebView+TextExtraction.swift` to the derived sources output file list.

* Source/WebKit/DerivedSources.make:

Add a DerivedSources make rule for `WKWebView+TextExtraction.swift` (copied from
`WKWebView+TextExtraction.swift.in`, in WebKitAdditions).

* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setTextExtractionEnabled:]):
(-[WKPreferences _textExtractionEnabled]):

Add an SPI flag for internal clients to enable or disable text extraction.

* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):

Add a new category that implements an ObjC selector, `-_requestTextExtraction:completionHandler:`,
which is invoked via `.perform(Selector((…)))` from `WKWebView+TextExtraction.swift`.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm:
(WebKit::prepareTextExtractionSupport):

Add a helper method to ensure that `libWebKitSwift` is loaded into memory; called when an internal
client enables text extraction via the new preference.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Add the new `WKWebView+TextExtraction.swift` derived source to WebKitSwift.

Canonical link: https://commits.webkit.org/274511@main
